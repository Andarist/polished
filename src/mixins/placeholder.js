// @flow
import messageHandlers from '../internalHelpers/_messageHandlers'

/**
 * CSS to style the selection psuedo-element.
 *
 * ***Deprecation Warning:*** *This mixin has been deprecated and will be removed in version 3.0.*
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...placeholder({'color': 'blue'})
 * }
 *
 * // styled-components usage
 * const div = styled.input`
 *    ${placeholder({'color': 'blue'})}
 * `
 *
 * // CSS as JS Output
 *
 * 'input': {
 *   '&:-moz-placeholder': {
 *     'color': 'blue',
 *   },
 *   '&:-ms-input-placeholder': {
 *     'color': 'blue',
 *   },
 *   '&::-moz-placeholder': {
 *     'color': 'blue',
 *   },
 *   '&::-webkit-input-placeholder': {
 *     'color': 'blue',
 *   },
 * },
 */
function placeholder(styles: Object, parent?: string = '&') {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      messageHandlers('mixins/placeholder.js', {
        // eslint-disable-next-line prefer-rest-params
        arrityCheck: { args: arguments, min: 1, max: 2 },
        typeChecks: [
          { param: styles, type: 'object' },
          { param: parent, type: 'string' },
        ],
      })
    ) {
      return {}
    }
  }

  return {
    [`${parent}::-webkit-input-placeholder`]: {
      ...styles,
    },
    [`${parent}:-moz-placeholder`]: {
      ...styles,
    },
    [`${parent}::-moz-placeholder`]: {
      ...styles,
    },
    [`${parent}:-ms-input-placeholder`]: {
      ...styles,
    },
  }
}

export default placeholder
