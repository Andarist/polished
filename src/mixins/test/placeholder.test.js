// @flow
import placeholder from '../placeholder'

describe('placeholder', () => {
  const styles = {
    color: 'blue',
  }

  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
      log: global.console.log,
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('should properly pass parent and styles object', () => {
    expect({
      ...placeholder('input', styles),
    }).toMatchSnapshot()
  })

  it('should properly default to & when not passed a parent', () => {
    expect({
      input: {
        ...placeholder(undefined, styles),
      },
    }).toMatchSnapshot()
  })

  it('should throw 2 errors when not passed any parameters', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    placeholder()
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string as its first parameter', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    placeholder(1, styles)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-object as its second parameter', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    placeholder('input', 'styles')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw a warning when passed more than 2 parameters', () => {
    placeholder('p', styles, true)
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })
})
