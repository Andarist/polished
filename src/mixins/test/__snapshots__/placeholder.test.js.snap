exports[`placeholder should properly default to & when not passed a parent 1`] = `
Object {
  "input": Object {
    "&:-moz-placeholder": Object {
      "color": "blue",
    },
    "&:-ms-input-placeholder": Object {
      "color": "blue",
    },
    "&::-moz-placeholder": Object {
      "color": "blue",
    },
    "&::-webkit-input-placeholder": Object {
      "color": "blue",
    },
  },
}
`;

exports[`placeholder should properly pass parent and styles object 1`] = `
Object {
  "input:-moz-placeholder": Object {
    "color": "blue",
  },
  "input:-ms-input-placeholder": Object {
    "color": "blue",
  },
  "input::-moz-placeholder": Object {
    "color": "blue",
  },
  "input::-webkit-input-placeholder": Object {
    "color": "blue",
  },
}
`;

exports[`placeholder should throw a warning when passed more than 2 parameters 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- mixins/placeholder.js --%c

placeholder %cwill be deprecated as of %cversion 3.0%c of ✨ polished. You should use the %c::placeholder pseudo-element%c instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#placeholder%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ WARNING ✨ ---- mixins/placeholder.js --%c

placeholder %cexpects a maximum of %c2 parameters%c. However, you passed %c3 parameters%c. This additional parameter was ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#placeholder%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`placeholder should throw an error when not passed any parameters 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/placeholder.js --%c

placeholder %crequires a %c2nd parameter%c(%cundefined%c: %cobject%c). However, you did not pass %cundefined%c.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#placeholder%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`placeholder should throw an error when passed a non-object as its second parameter 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/placeholder.js --%c

placeholder %cexpects a %c2nd parameter%c(%cundefined%c: %cobject%c). However, you passed (%c\'styles\'%c:%cstring%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#placeholder%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`placeholder should throw an error when passed a non-string as its first parameter 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/placeholder.js --%c

placeholder %cexpects a %c1st parameter%c(%cundefined%c: %cstring%c). However, you passed (%c\'1\'%c:%cnumber%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#placeholder%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
